make/Images.gmk

```diff
@@ -122,32 +128,36 @@ define CreateCDSArchive
   $1_$2_DUMP_EXTRA_ARG := $(if $(filter _nocoops, $2),-XX:-UseCompressedOops,)
   $1_$2_DUMP_TYPE      := $(if $(filter _nocoops, $2),-NOCOOPS,)
 
+  # Only G1 supports dumping the shared heap, so explicitly use G1
+  # it if the JVM supports it. (Note: the default GC with zero is SerialGC)
+  $1_$2_CDS_DUMP_FLAGS := $(CDS_DUMP_FLAGS) $(if $(filter g1gc, $(JVM_FEATURES_$1)),-XX:+UseG1GC)
```

Comment:
I think it should be e.g. (untested):

```suggestion
  $1_$2_CDS_DUMP_FLAGS := $(CDS_DUMP_FLAGS)
  ifeq ($(call check-jvm-feature, g1gc), true)
     $1_$2_CDS_DUMP_FLAGS += -XX:+UseG1GC
  endif
```

Reply:
That doesn't work because `check-jvm-feature` requires `JVM_VARIANT` to be set, but `CreateCDSArchive` is not called in a context where  `JVM_VARIANT` is set. ( `JVM_VARIANT` is set only in a few specific places in Main.gmk, etc).

One option is to change the foreach loop a few lines below to:

```
  $(foreach JVM_VARIANT, $(JVM_VARIANTS), \
    $(eval $(call CreateCDSArchive,)) \
  )
```

But I've not seen `JVM_VARIANT` being used this way, so I am a little hesitant about doing it.

Reply:
Is the CDS archive dumped in a JVM variant specific way? If so, then it would make sense to dump it for each configured variant.

Reply:
In that case, the loop you proposed make a lot of sense and I think you should add it.

Reply:
Oh, the loop is already there, it's jus the naming of the variable. I think we should adjust that so the check-jvm-feature macro can be used.

Reply:
@erikj79 I couldn't get it to work. I believe the following is evaluated only once, so `$(JVM_VARIANT)` is bound to the value when HotspotCommon.gmk is loaded:

```
check-jvm-feature = \
  $(strip \
    $(if $(filter-out $(VALID_JVM_FEATURES), $1), \
      $(error Internal error: Invalid feature tested: $1)) \
    $(if $(filter $1, $(JVM_FEATURES_$(JVM_VARIANT))), true, false))
```

So I would need to run a submake in order to get this function to behave differently for each variant. That's going to be complicated in order to support multiple JVM variants per build.

I heard that we are going to remove multiple variant support (?), so maybe this is not worth it.

Reply:
The suggested code should be something like this, note the double `$$` in the call.

```suggestion
  $1_$2_CDS_DUMP_FLAGS := $$(CDS_DUMP_FLAGS)
  ifeq ($$(call check-jvm-feature, g1gc), true)
    $1_$2_CDS_DUMP_FLAGS += -XX:+UseG1GC
  endif
```

Reply:
I just realized that check-jvm-feature is only defined in HotspotCommon.gmk, which isn't included here (and can't really be either without modifications). Sorry to lead you on a wild goose chase. Unless Magnus can think of a reasonable way of exposing that macro here, I think your current solution is good enough.

Reply:
Yes, it is good enough. That macro is supposed to work only in the context of the hotspot building.